solver/src/liquidity/offchain_orderbook.rs

```diff
@@ -47,6 +51,194 @@ pub fn normalize_limit_order(order: Order, native_token: WETH9) -> LimitOrder {
 
 impl SettlementHandling<LimitOrder> for OrderSettlementHandler {
     fn encode(&self, executed_amount: U256, encoder: &mut SettlementEncoder) -> Result<()> {
-        encoder.add_trade(self.order.clone(), executed_amount)
+        if self.order.order_creation.buy_token == BUY_ETH_ADDRESS {
+            let interaction = compute_unwrap_interaction(
+                encoder.clearing_prices(),
+                &self.order,
+                executed_amount,
+                self.native_token.clone(),
+            )?;
+
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+
+            encoder.add_unwrap(interaction);
+        } else {
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn compute_unwrap_interaction(
+    clearing_prices: &HashMap<H160, U256>,
+    order: &Order,
+    executed_amount: U256,
+    weth: WETH9,
+) -> Result<UnwrapWethInteraction> {
+    let sell_price = if let Some(&price) = clearing_prices.get(&order.order_creation.sell_token) {
+        price
+    } else {
+        return Err(anyhow!("sell price not available"));
+    };
```

Comment:
nit: This can be simplified with `ok_or_else` to take advantage of "try-sugar":
```suggestion
    let sell_price = clearing_prices.get(&order.order_creation.sell_token)
        .ok_or_else(|| anyhow!("sell price not available"))?;
```

Same applies to other `Option::None` handing :point_down: